### PR TITLE
Add Execution Attributes Overrides for Client and Request Configurations

### PR DESCRIPTION
**Pull request by ndejaco2**
## Description
This change adds getters/setters for execution attributes on the client and request override configuration objects in order to allow these attributes to be added prior to a request being made. The additional execution attributes are added to the collection at the time of a request.

## Motivation and Context
The motivation for this change is to allow a request to pass a custom object override instead of relying on thread local storage for passing data to a client interceptor. 

## Testing
Unit tests added for getters and setters. 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License
- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java [hunk 1]
```diff
@@ -46,6 +47,7 @@
     private final Duration apiCallAttemptTimeout;
     private final Signer signer;
     private final List<MetricPublisher> metricPublishers;
+    private final Map<ExecutionAttribute<?>, Object> executionAttributes;
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java:50]: We should use [`AttributeMap`](https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/AttributeMap.java)

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java [hunk 2]
```diff
@@ -439,6 +451,22 @@ default Builder retryPolicy(RetryMode retryMode) {
...
+         * @return This object for method chaining.
+         */
+        Builder executionAttributes(Map<ExecutionAttribute<?>, Object> executionAttributes);
+
+        /**
+         * Add an execution attribute to the existing collection of execution attributes.
+         * @param attribute The execution attribute object
+         * @param value The value of the execution attribute.
+         */
+        Builder addExecutionAttribute(ExecutionAttribute attribute, Object value);
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java:467]: - Let's use the `put` verb here. Typically `add` is used if you're adding to a existing set under the given key.
 - Can we make this typed? i.e. `addExecutionAttribute(ExecutionAttribute<T> attribute, T value)`

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java [hunk 3]
```diff
@@ -439,6 +451,22 @@ default Builder retryPolicy(RetryMode retryMode) {
         Builder addMetricPublisher(MetricPublisher metricPublisher);
 
         List<MetricPublisher> metricPublishers();
+
+        /**
+         * Sets the additional execution attributes collection for this client.
+         * @param executionAttributes Execution attributes map for this client.
+         * @return This object for method chaining.
+         */
+        Builder executionAttributes(Map<ExecutionAttribute<?>, Object> executionAttributes);
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java:460]: Let's use `Map<ExecutionAttribute<?>, ?>`

File: core/aws-core/src/main/java/software/amazon/awssdk/awscore/client/handler/AwsClientHandlerUtils.java [hunk 4]
```diff
@@ -100,6 +101,11 @@ private AwsClientHandlerUtils() {
             .putAttribute(SdkInternalExecutionAttribute.DISABLE_HOST_PREFIX_INJECTION,
                           clientConfig.option(SdkAdvancedClientOption.DISABLE_HOST_PREFIX_INJECTION));
 
+        Optional.ofNullable(clientConfig.option(SdkClientOption.EXECUTION_ATTRIBUTES)).ifPresent(clientOverrideConfiguration ->
+                executionAttributes.putAllAttributes(clientOverrideConfiguration));
+        originalRequest.overrideConfiguration().ifPresent(requestOverrideConfiguration ->
+                executionAttributes.putAllAttributes(requestOverrideConfiguration.executionAttributes()));
+
```
Comment (dagnir) [inline core/aws-core/src/main/java/software/amazon/awssdk/awscore/client/handler/AwsClientHandlerUtils.java:108]: Let's simplify this by using [`AttributeMap#merge()`](https://github.com/aws/aws-sdk-java-v2/blob/1df4bd27fce7f9eb8c8670b0417057b63bba0ae3/utils/src/main/java/software/amazon/awssdk/utils/AttributeMap.java#L63-L74) to merge the request level and client level execution attribute overrides, giving the request level attributes precedence.

**Review by dagnir — CHANGES_REQUESTED** [review]
Can we add some tests to ensure that the client interceptors see these override values? See this test class for an example https://github.com/aws/aws-sdk-java-v2/blob/1df4bd27fce7f9eb8c8670b0417057b63bba0ae3/test/codegen-generated-classes-test/src/test/java/software/amazon/awssdk/services/ExecutionAttributesTest.java#L23

**Comment by dagnir** [comment]
Should we consider plumbing the execution attributes all the way to the response/exception returned from the API call? I don't think this is a blocker for the PR as it can be added later, but want to see if you think there's a benefit there, especially as a user of this feature.

**Comment by dagnir** [comment]
BTW the pattern used on the builders is identical to `ClientOverrideConfiguration#putAdvancedOption()`. Can you make sure to follow the same pattern? My previous comments should still apply

https://github.com/aws/aws-sdk-java-v2/blob/1df4bd27fce7f9eb8c8670b0417057b63bba0ae3/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java#L322-L330

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java (hunk 3 above)
Comment (ndejaco2) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java:460]: This wont work when calling putExecutionAttribute the type generic will not be able to be converted to ?. "Error:(572, 62) java: incompatible types: T cannot be converted to capture#1 of ?"

File: core/aws-core/src/main/java/software/amazon/awssdk/awscore/client/handler/AwsClientHandlerUtils.java [hunk 5]
```diff
@@ -100,6 +101,17 @@ private AwsClientHandlerUtils() {
             .putAttribute(SdkInternalExecutionAttribute.DISABLE_HOST_PREFIX_INJECTION,
                           clientConfig.option(SdkAdvancedClientOption.DISABLE_HOST_PREFIX_INJECTION));
 
+        ExecutionAttributes clientOverrideExecutionAttributes = clientConfig.option(SdkClientOption.EXECUTION_ATTRIBUTES);
+        if (clientOverrideExecutionAttributes != null) {
+            executionAttributes = clientOverrideExecutionAttributes.merge(executionAttributes);
+        }
+
+        Optional<RequestOverrideConfiguration> requestOverrideConfiguration =
+                (Optional<RequestOverrideConfiguration>) originalRequest.overrideConfiguration();
```
Comment (dagnir) [inline core/aws-core/src/main/java/software/amazon/awssdk/awscore/client/handler/AwsClientHandlerUtils.java:110]: Hmm I don't like the cast. Can we just keep the type as `Optional<? extends RequestOverrideConfiguration>`?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java [hunk 6]
```diff
@@ -138,6 +142,16 @@ protected RequestOverrideConfiguration(Builder<?> builder) {
...
     }
 
+    /**
+     * Returns the additional execution attributes to be added to this request.
+     * This collection of attributes is added in addition to the attributes set on the client.
+     * An attribute value added on the client within the collection of attributes is superseded by an
+     * attribute value added on the request.
+     */
+    public ExecutionAttributes executionAttributes() {
+        return executionAttributes;
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java:152]: `RequestOverrideConfiguration` should be immutable; can we create an immutable view of `executionAttributes` to return?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttributes.java [hunk 7]
```diff
@@ -48,11 +66,76 @@
...
+     */
+    public <U> ExecutionAttributes putAllAttributes(Map<ExecutionAttribute<?>, ?> executionAttributes) {
+        this.attributes.putAll(executionAttributes);
+        return this;
+    }
+
+    /**
+    * Merge attributes of a lower precedence into the current higher precedence current collection.
+    */
+    public <U> ExecutionAttributes merge(ExecutionAttributes lowerPrecedenceExecutionAttributes) {
```
Comment (dagnir) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttributes.java:80]: Please add a test for this

Comment (ndejaco2) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttributes.java:80]: this is tested by asyncClient_requestOverrideExecutionAttributesHavePrecedence test in ExecutionAttributesTest

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java (hunk 6 above)
Comment (ndejaco2) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/RequestOverrideConfiguration.java:152]: Like a UnmodifiableMap?

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttributes.java (hunk 7 above)
Comment (ndejaco2) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttributes.java:80]: added an additional test for this which required some additional changes.

Comment (ndejaco2) [reply core/sdk-core/src/main/java/software/amazon/awssdk/core/interceptor/ExecutionAttributes.java:80]: I will add additional tests

**Comment by ndejaco2** [comment]
> Should we consider plumbing the execution attributes all the way to the response/exception returned from the API call? I don't think this is a blocker for the PR as it can be added later, but want to see if you think there's a benefit there, especially as a user of this feature.

This is already done in the callbacks for interceptors. However, there is one issue which I opened up today: https://github.com/aws/aws-sdk-java-v2/issues/2367

**Comment by ndejaco2** [comment]
> BTW the pattern used on the builders is identical to `ClientOverrideConfiguration#putAdvancedOption()`. Can you make sure to follow the same pattern? My previous comments should still apply
> 
> https://github.com/aws/aws-sdk-java-v2/blob/1df4bd27fce7f9eb8c8670b0417057b63bba0ae3/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java#L322-L330

Resolved

**Comment by ndejaco2** [comment]
> > BTW the pattern used on the builders is identical to `ClientOverrideConfiguration#putAdvancedOption()`. Can you make sure to follow the same pattern? My previous comments should still apply
> > https://github.com/aws/aws-sdk-java-v2/blob/1df4bd27fce7f9eb8c8670b0417057b63bba0ae3/core/sdk-core/src/main/java/software/amazon/awssdk/core/client/config/ClientOverrideConfiguration.java#L322-L330
> 
> Resolved

**Comment by dagnir** [comment]
LGTM! Thank you for the PR :). Can you add a changelog entry and I can give this a ship-it?

**Comment by ndejaco2** [comment]
> LGTM! Thank you for the PR :). Can you add a changelog entry and I can give this a ship-it?
> 
> Details here: https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md#pull-request-readiness

Thanks for your help. Committed it now

**Review by dagnir — APPROVED** [review]

**Pull request merged**